### PR TITLE
Cached metrics

### PR DESCRIPTION
**Pull request by shcheklein**
* Fixes https://github.com/iterative/dvc/issues/1590
* Adds ability to run a specific test from a scope:

```
python -m tests tests/test_run.py:TestCmdRunCliMetrics
```

or 

```
python -m tests tests/test_run.py:TestCmdRunCliMetrics:test_not_cached
```
 
* Fixes some warnings from PyCharm's linter
* Fixes a few error messages and edge case tests

### CODE REVIEW

File: tests/__main__.py [hunk 1]
```diff
@@ -10,8 +12,14 @@
...
+if len(sys.argv) == 1:
+    scope = "--all-modules"
+else:
+    scope = " ".join(sys.argv[1:])
+
 cmd = (
     "nosetests -v --processes=-1 --process-timeout=500 --cover-inclusive "
-    "--cover-erase --cover-package=dvc --with-coverage --all-modules --with-flaky"
+    "--cover-erase --cover-package=dvc --with-coverage --with-flaky "
+    "{scope} ".format(scope=scope)
```
Comment (mroutis) [inline tests/__main__.py]: :zap: This was the only reason I was using `pytest` instead of `python -m tests`!

Also, can we change the name to `extra-args` or something like that? I'm thinking of using `python -m tests -x` or `python -m tests -s tests/test_something.py`

File: dvc/exceptions.py [hunk 2]
```diff
@@ -11,7 +11,7 @@ class DvcException(Exception):
     """Base class for all dvc exceptions.
 
     Args:
-        msg (str): message for this exception.
+        msg (unicode): message for this exception.
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/exceptions.py]: Yep! This exactly why it should be `unicode` not `str`. Otherwise PyCharm's linter complains about type mismatch.

File: dvc/logger.py [hunk 3]
```diff
@@ -230,13 +231,13 @@ def _walk_exc(exc):
             tb_list.insert(0, str(exc.cause_tb))
         exc = exc.cause
 
-    return (exc_list, tb_list)
+    return exc_list, tb_list
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/logger.py]: yep, good question. Do we have it in code style, @efiop ? I think PyCharm is set by default this way. I don't have a strong preference, to be honest, but like w/o them :)

File: dvc/project.py [hunk 4]
```diff
@@ -370,10 +370,11 @@ def unprotect(self, path):
     def run(
         self,
         cmd=None,
-        deps=[],
-        outs=[],
-        outs_no_cache=[],
-        metrics_no_cache=[],
+        deps=None,
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/project.py]: 👍

File: dvc/project.py [hunk 5]
```diff
@@ -383,6 +384,17 @@ def run(
     ):
         from dvc.stage import Stage
 
+        if outs is None:
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/project.py]: yep, I hate all these ifs, some refactoring is required around this method. `outs = outs or []` has different semantics unfortunately - it will return the second `[]` in this case `[] or []` as far as I remember.

File: tests/__main__.py (hunk 1 above)
Comment (shcheklein) [reply tests/__main__.py]: I'm not I understand your suggestion, could you give a specific example why would we need `-x` or `-s`?

**Comment by shcheklein** [comment]
Thanks, @mroutis ! 👍

File: tests/__main__.py (hunk 1 above)
Comment (mroutis) [reply tests/__main__.py]:
```
-x, --stop            Stop running tests after the first error or failure
-s, --nocapture       Don't capture stdout (any stdout output will be printed immediately) 
```

`-s` to make it work with `import ipdb; ipdb.set_trace()` (sometimes I need a step by step debugger to understand what's going on, I find it more broad to explore the codebase than `print`ing some lines)

Comment (shcheklein) [reply tests/__main__.py]: probably `scope` is the wrong var name, I think it should work if you pass any arguments, (you will have to specify the scope though) - `python -m tests -s -x tests` - is it good enough, @mroutis ? what do you think?

File: dvc/project.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mroutis) [reply dvc/project.py]: It isn't the intended, `only if outs is empty, initialize its value`?

```python
outs = None                                                                   
outs = outs or []                                                             
# >>> outs
# []

outs = outs or ['x', 'y']
# >>> outs
# ['x', 'y']

outs = outs or []
# >>> outs
# ['x', 'y']
```

The only difference is that the `val = val or new_val` checks for "falsy" values instead of just `None` (so `outs = False or "x"` will assign `"x"`, even if `False` is a valid value).

File: tests/__main__.py (hunk 1 above)
Comment (shcheklein) [reply tests/__main__.py]: Btw, if you use debugger I would strongly recommend using PyCharm - you can run any test with one click and have really nice debugger with a lot nice UI features.

File: dvc/command/run.py [hunk 6]
```diff
@@ -95,9 +96,16 @@ def add_parser(subparsers, parent_parser):
         "--outs-no-cache",
         action="append",
         default=[],
-        help="Declare output regular file or directory "
+        help="Declare output data file or directory "
```
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/command/run.py]: yes, good point ... I'll update it more carefully

File: tests/__main__.py (hunk 1 above)
Comment (mroutis) [reply tests/__main__.py]: Thanks for the recommendation, I'll definetly take a look into it, maybe is time to bury the _good ol' Vim_.

`python -m tests -s -x tests` works for me :)

Would you mind editing the README.md to add information about how to run a single test?

File: dvc/project.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/project.py]: My intuition to is to keep it as is - don't optimize by changing semantics just because it looks better. It breaks the protocol, it basically converts any effectively falsey value into an empty array. I think it's not a good practice. Let's imagine someone by mistake passes `outs=False`. Should we support this or fail? Maybe in this specific case even `outs or []` semantics is what we need, but I would double check.

File: tests/__main__.py (hunk 1 above)
Comment (shcheklein) [reply tests/__main__.py]: Totally, will update it!

File: dvc/logger.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (efiop) [reply dvc/logger.py]: @shcheklein we don't. black doesn't seem to format it either. Let's leave it as is, maybe pylint will catch it later.

[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/logger.py]: 👍 put them back for now. Just for the reference - http://google.github.io/styleguide/pyguide.html#33-parentheses . We should embrace one single code style as soon as possible. Like I said before I don't care that much specifically about tuples - we just should have a single guide.

File: dvc/command/run.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (shcheklein) [reply dvc/command/run.py]: updated it a little bit, check also the second line. I thinks it's good enough for now.

File: tests/__main__.py (hunk 1 above)
Comment (shcheklein) [reply tests/__main__.py]: Will be updating the contributing guide, not README.md
